store/bench_test.go

```diff
@@ -0,0 +1,378 @@
+		if len(sampledPercentages) > 0 {
+			for _, p := range sampledPercentages {
+				name := fmt.Sprintf("r-%s-%d-%d-%d-%d", dbType, p.has, p.get, p.set, p.delete)
+				benchmarks = append(benchmarks, benchmark{name: name, percentages: p, dbType: dbType, counts: counts{}})
+			}
+		} else if len(sampledCounts) > 0 {
+			for _, c := range sampledCounts {
+				name := fmt.Sprintf("d-%s-%d-%d-%d-%d", dbType, c.has, c.get, c.set, c.delete)
+				benchmarks = append(benchmarks, benchmark{name: name, percentages: percentages{}, dbType: dbType, counts: c})
+			}
+		}
+	}
+	return benchmarks
+}
+
+type store interface {
+	Has(key []byte) bool
+	Get(key []byte) []byte
+	Set(key []byte, value []byte)
+	Delete(key []byte)
+	Commit() types.CommitID
+}
+
+type storeV2 struct {
+	*storev2.Store
+	storev2types.KVStore
+}
+
+func simpleStoreConfig() (storev2.StoreConfig, error) {
+	opts := storev2.DefaultStoreConfig()
+	err := opts.RegisterSubstore(skey_1.Name(), types.StoreTypePersistent)
+	if err != nil {
+		return storev2.StoreConfig{}, err
+	}
+	return opts, nil
+}
+
+func sampleOperation(p percentages) string {
+	ops := []string{"Has", "Get", "Set", "Delete"}
+	thresholds := []int{p.has, p.has + p.get, p.has + p.get + p.set}
+	r := rand.Intn(100)
+	for i := 0; i < len(thresholds); i++ {
+		if r < thresholds[i] {
+			return ops[i]
+		}
+	}
+	return ops[3]
+}
+
+func runRandomizedOperations(b *testing.B, s store, totalOpsCount int, p percentages) {
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		for j := 0; j < totalOpsCount; j++ {
+			b.StopTimer()
+			op := sampleOperation(p)
+			b.StartTimer()
+
+			switch op {
+			case "Has":
+				s.Has(randBytes(12))
+			case "Get":
+				s.Get(randBytes(12))
+			case "Set":
+				s.Set(randBytes(12), randBytes(50))
+			case "Delete":
+				s.Delete(randBytes(12))
+			}
+			if j%200 == 0 || j == totalOpsCount-1 {
+				s.Commit()
+			}
+		}
+	}
+}
+
+func prepareValues() [][]byte {
+	var data [][]byte
+	for i := 0; i < 5000; i++ {
+		data = append(data, randBytes(50))
+	}
+	return data
+}
+
+func createKey(i int) []byte {
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, uint64(math.Sin(float64(i))*100000))
+	return b
+}
+
+func runDeterministicOperations(b *testing.B, s store, values [][]byte, c counts) {
+	counts := []int{c.has, c.get, c.set, c.delete}
+	sort.Ints(counts)
+	step := counts[len(counts)-1]
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		idx := i * step
+
+		b.StopTimer()
+		if idx >= len(values) {
+			for j := len(values); j < (idx + step); j++ {
+				values = append(values, randBytes(50))
+			}
+		}
+
+		b.StartTimer()
+		for j := 0; j < c.set; j++ {
+			key := createKey(idx + j)
+			s.Set(key, values[idx+j])
+		}
+		for j := 0; j < c.has; j++ {
+			key := createKey(idx + j)
+			s.Has(key)
+		}
+		for j := 0; j < c.get; j++ {
+			key := createKey(idx + j)
+			s.Get(key)
+		}
+		for j := 0; j < c.delete; j++ {
+			key := createKey(idx + j)
+			s.Delete(key)
+		}
+		s.Commit()
```

Comment:
Here you're measuring an entire transaction of arbitrarily many operations in each benchmark loop. This means each call of this function with different `counts` are not comparable. Is that your intent?